src/Storages/MergeTree/MergeTreeDataPartWriterOnDisk.cpp

```diff
@@ -10,13 +11,24 @@ namespace ErrorCodes
     extern const int LOGICAL_ERROR;
 }
 
-void MergeTreeDataPartWriterOnDisk::Stream::finalize()
+void MergeTreeDataPartWriterOnDisk::Stream::preFinalize()
 {
     compressed.next();
     /// 'compressed_buf' doesn't call next() on underlying buffer ('plain_hashing'). We should do it manually.
     plain_hashing.next();
     marks.next();
 
+    plain_file->preFinalize();
+    marks_file->preFinalize();
+
+    is_prefinalized = true;
+}
+
+void MergeTreeDataPartWriterOnDisk::Stream::finalize()
+{
+    if (!is_prefinalized)
```

Comment:
Yes, I can only make `preFinalize` an optional optimisation, cause otherwise I would need to fix all other write buffer usages. Btw we don't even call `finalize()` all the time now, so that it may be called from the destructor.